src/codegen/encoding/mod.rs

```diff
@@ -284,40 +761,21 @@ fn calculate_array_size<T: AbiEncoding>(
             vartab,
             cfg,
         );
-        size_var
+        Expression::Variable(LOC, U32, size_var)
     };
 
-    // Each dynamic dimension size occupies 4 bytes in the buffer
-    let dyn_dims = dims.iter().filter(|d| **d == ArrayLength::Dynamic).count();
-
-    if dyn_dims > 0 && !encoder.is_packed() {
-        cfg.add(
-            vartab,
-            Instr::Set {
-                loc: Loc::Codegen,
-                res: size_var,
-                expr: Expression::Add(
-                    Loc::Codegen,
-                    Type::Uint(32),
-                    false,
-                    Box::new(Expression::Variable(Loc::Codegen, Type::Uint(32), size_var)),
-                    Box::new(Expression::NumberLiteral(
-                        Loc::Codegen,
-                        Type::Uint(32),
-                        BigInt::from(4 * dyn_dims),
-                    )),
-                ),
-            },
-        );
+    if !encoder.is_packed() && matches!(&dims.last().unwrap(), ArrayLength::Dynamic) {
```

Comment:
I read the code of the function as follows: If we have `direct_assessment`, then
- We have either 0 dynamic dimensions
- Or exactly one, which is the last one
- We don't recursively walk `get_expr_size`, so we won't account for the encoded length size automatically

Thus follows: If we are not in packed encoder *and* the last dimension is dynamic (since direct assessment can be true if we don't have dynamic dimensions at all), we need to account for the size of the encoded length.

Hence, I am only checking for the last one to be dynamic, because only in that case this code path can be reached. Or am I getting this wrong?

This reminds me that it would deserve a comment or two though.

Reply:
This condition is executed regardless of `direct_assessment`'s value, so you need to consider all cases in which dynamic dimensions are available.

Reply:
> This condition is executed regardless of direct_assessment's value

I am aware of that. However, `calculate_complex_array_size` calls `get_expr_size` recursively. Any size length information should be handled there. Except when we are in direct assessment. Then we need to handle the size length information manually. But in direct assessment, only the last dimension can be dynamic. Hence this check. Can you provide me with a test case that fails in the situation you describe?

Reply:
```solidity
contract PTokenFactory {
    function strConcat(uint16[][3][] memory vec) public pure returns (bytes memory){
        bytes b = abi.encode(vec);
        return b;
    }
}
```

If you emit the CFG for this case, you'll see that the space necessary to save both dynamic dimensions is not accounted for while we calculate the size of the array.

Later, when the actual buffer is allocated, we have this instruction:

`ty:bytes %abi_encoded.temp.7 = (alloc bytes len (%array_bytes_size_0.temp.3 + uint32 4))`. It accounts only one of the dynamic dimensions. We should have a `+ uint32 8` here.

PS: I have run this command: ` ./target/debug/solang compile --target solana ../solang/examples/test.sol --emit cfg`

Reply:
Hmm, this crashes the compiler :neutral_face: It would be nice to have a functioning test for this...

However I'm still convinced the accounting for the encoded size must happen inside `calculate_complex_array_size`, where the actually array size calculation is happening. We need the actual array size value to calculate it's width, and this is the (only) place where it gets calculated?

Reply:
I'm confused by the `calculate_complex_array_size` function. Why do we recursively write for loops into our CFG, but then only calculate the array length for the dimension `0`?

Reply:
> Hmm, this crashes the compiler 😐 It would be nice to have a functioning test for this...

If you do `--emit `, you can see the generated code and follow the CFG instructions to see what I'm telling you. There are many bugs with multidimensional arrays that needs fixing, however the encoder was designed to handle all cases already.

> However I'm still convinced the accounting for the encoded size must happen inside `calculate_complex_array_size`, where the actually array size calculation is happening.

No, it does not. The `calculate_complex_array_size ` only calculates the size of each element in the array. It does not account for the space needed to save the dimension. This is how I designed the function.

> I'm confused by the calculate_complex_array_size function. Why do we recursively write loops into our CFG, but then only calculate the array length for the dimension 0? 

if your vector is like `my_struct[][3][]`,  `calculate_complex_array_size ` will implement this only:

```
uint32 size = 0;
for (int i=0; i<my_struct.length; i++) {
    for(int j=0; j<3; j++) {
        for(int k=0; k<my_struct[i][j].length; k++) {
            size += size_of(my_struct[i][j][k]);
        }
    }
}
```

I'm gonna fix the bug in the example and get back to you. Meanwhile, you can check the CFG by doing `--emit cfg`, because that still works.

Reply:
> If you do --emit , you can see the generated code and follow the CFG instructions to see what I'm telling you. 

I know. I pointed this out because now I'm working around a problem that we can not even properly test. As a side note, in the future we should avoid having to maintain such code. If we can not test some functionality properly, because it depends on other things that are missing, it is a strong hint that it would be best to avoid implementing such functionality, or fix / add whats needed first.

> I'm gonna fix the bug in the example and get back to you. 

This is a great idea. Thanks. It makes no sense to try to work with such such high-dimensional arrays in encoding, when they are broken anyways to a point it is not even possible to write a test. Note that another option (as an immediate fix) here would be to just forbid array dimensions > 2 in sema (and assert that here). This would allow to simplify this code and fix multidimensional arrays everywhere for good later on. Acutally thinking about it, until we haven't fixed such arrays for good and have good test coverage for them, it would make sense to forbid this in sema anyways.

> The calculate_complex_array_size only calculates the size of each element in the array. It does not account for the space needed to save the dimension. This is how I designed the function.

Fair enough, but I don't see any other solution to get this going for SCALE encoding than to account for the encoded length here. Also this would then be consistent to what is being done in `encode_complex_array_size`. What do you think?

Reply:
Ok so apparently, this is not something that is trivial to fix. So we should really to stop worrying about arrays with >2 dimensions here until we have everything to 
1. Properly test this
2. Properly announce the feature ready to use.

I'll wait for #1132 to go in nonetheless, but then just forbid more than 2 array dimensions in sema. I'd like to go forward with this PR.

Reply:
> then just forbid more than 2 array dimensions in sema.

Multidimensional fixed arrays work perfectly. We have tests for them in the encoder/decoder. I would not disable that right now, as we won't be able to detect regressions in the tests we already have.

Reply:
https://github.com/hyperledger/solang/blob/468e4d6a59794e0ac4a74001986f600bec0a6d68/tests/solana_tests/abi_encode.rs#L581

https://github.com/hyperledger/solang/blob/468e4d6a59794e0ac4a74001986f600bec0a6d68/tests/solana_tests/abi_encode.rs#L581

Reply:
#1132 has been merged.